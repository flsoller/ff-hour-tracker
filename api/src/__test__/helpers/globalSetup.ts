/* eslint-disable @typescript-eslint/no-var-requires */
const util = require('util');
const exec = util.promisify(require('child_process').exec);

/**
 * Run migrations to initialize db
 */
module.exports = async () => {
  await exec(`yarn prisma migrate reset --force`);
};

export {};
