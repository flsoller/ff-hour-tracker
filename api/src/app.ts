// Package imports
import express, { Application } from 'express';
import cors from 'cors';
import dotenv from 'dotenv';

// Middleware imports
import errorHandler from './middleware/errorHandler';

// Controller imports
import organizations from './controllers/Organizations';
import members from './controllers/Members';
import activities from './controllers/Activities';
import timelog from './controllers/TimeLogs';

// Load env configuration
dotenv.config();

// Initalize express
const app: Application = express();

// Allow cors for development
if (process.env.NODE_ENV === 'development') {
  app.use(cors());
}

// JSON body-parser middleware
app.use(express.json());

// Mount routes
const API_V0 = '/api/v0';
app.use(`${API_V0}/organizations`, organizations);
app.use(`${API_V0}/members`, members);
app.use(`${API_V0}/activities`, activities);
app.use(`${API_V0}/timelog`, timelog);

// Error handling middleware
app.use(errorHandler);

export default app;
