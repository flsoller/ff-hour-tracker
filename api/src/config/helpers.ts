import sequelize from './db';

export const connectDb = async (): Promise<void> => {
  try {
    await sequelize.authenticate();
    console.log('Connection has been established successfully.');
  } catch (error) {
    console.error('Unable to connect to the database:', error);
  }
};
