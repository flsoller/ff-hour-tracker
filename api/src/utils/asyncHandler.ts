import { NextFunction, Request, Response } from 'express';

// asyncHandler to avoid repeating try/catch on every async/await controller method.
const asyncHandler =
  (fn: any) => (req: Request, res: Response, next: NextFunction) =>
    Promise.resolve(fn(req, res, next)).catch(next);

export default asyncHandler;
