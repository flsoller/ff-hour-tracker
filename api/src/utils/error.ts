export class ErrorResponse extends Error {
  statusCode: number;
  additionalInfo = {};

  constructor(message: string, statusCode: number, additionalInfo = {}) {
    super(message);
    this.statusCode = statusCode || 500;
    this.additionalInfo = { ...additionalInfo };
  }
}
