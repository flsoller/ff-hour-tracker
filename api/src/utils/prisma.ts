import { PrismaClient } from '@prisma/client';

// Declare prisma on global object to avoid implicit any error
declare global {
  // eslint-disable-next-line no-var
  var prisma: PrismaClient | undefined;
}

/**
 * Instantiate a single instance of prisma
 */
export const prisma =
  global.prisma ||
  new PrismaClient({
    log: ['query'],
  });

if (process.env.NODE_ENV !== 'production') global.prisma = prisma;
