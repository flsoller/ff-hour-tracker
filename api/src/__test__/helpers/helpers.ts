import { prisma } from '../../utils/prisma';
import { registerUser } from '../../services/auth';
import { IRegisterUser } from '@hour-tracker/core-types/api/auth';
import supertest from 'supertest';
import { Role } from '@prisma/client';

async function createOrganizations() {
  const organization = await prisma.organization.create({
    data: {
      name: 'organization',
      description: 'organization description',
    },
  });

  const otherOrganization = await prisma.organization.create({
    data: {
      name: 'otherOrganization',
      description: 'otherOrganization description',
    },
  });

  return [organization, otherOrganization];
}

async function createMemberForOrganization(
  orgId: string,
  emailAddress = 'test@mail.com',
  firstName = 'Test',
  lastName = 'User',
) {
  return prisma.member.create({
    data: {
      firstName,
      lastName,
      emailAddress,
      orgId,
    },
  });
}

async function createUserForOrganization(orgId: string, options = {}) {
  const params: IRegisterUser = {
    emailAddress: 'testuser@db.com',
    password: '12345',
    name: 'Mr. Test',
    orgId,
    role: Role.USER,
    ...options,
  };

  return registerUser(params);
}

async function loginTestUser(
  app: Express.Application,
  emailAddress: string,
): Promise<string> {
  return supertest(app)
    .post('/api/v0/auth/signin')
    .send({
      emailAddress,
      password: '12345',
    })
    .then((res) => res.body.accessToken);
}

export {
  createOrganizations,
  createMemberForOrganization,
  createUserForOrganization,
  loginTestUser,
};
