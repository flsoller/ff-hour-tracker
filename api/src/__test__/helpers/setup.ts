import { prisma } from '../../utils/prisma';
jest.setTimeout(10000);

beforeEach(async () => {
  await prisma.$transaction([
    prisma.$executeRaw`TRUNCATE organizations CASCADE;`,
    prisma.$executeRaw`TRUNCATE members CASCADE;`,
    prisma.$executeRaw`TRUNCATE timelogs CASCADE;`,
    prisma.$executeRaw`TRUNCATE activity_types CASCADE;`,
    prisma.$executeRaw`TRUNCATE users CASCADE;`,
  ]);
});
