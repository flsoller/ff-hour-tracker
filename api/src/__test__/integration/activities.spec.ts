import app from '../../app';
import supertest from 'supertest';
import { prisma } from '../../utils/prisma';
import { createOrganizations } from '../helpers/helpers';
import { Organization } from '@prisma/client';

describe('activities controller', () => {
  const ROUTE = '/api/v0/activities';
  let organization: Organization;
  let otherOrganization: Organization;

  beforeEach(async () => {
    [organization, otherOrganization] = await createOrganizations();
  });

  describe('createActivity', () => {
    it('should create an activity for an organization', async () => {
      const res = await supertest(app).post(ROUTE).send({
        activityName: 'Test Activity',
        activityDesc: 'This is a nice activity',
        organizationId: organization.id,
      });

      expect(res.status).toBe(201);
      expect(res.body).toMatchSnapshot({
        id: expect.any(String),
      });
      expect(await prisma.activityType.count()).toBe(1);
    });

    it('should fail when required fields are missing', async () => {
      const res = await supertest(app).post(ROUTE).send({
        activityDesc: 'This is a nice activity',
        organizationId: organization.id,
      });

      expect(res.status).toBe(400);
      expect(res.body).toStrictEqual({
        error: 'MissingInformation',
        additionalInfo: {},
      });
    });

    it('should fail when organization does not exist', async () => {
      const res = await supertest(app).post(ROUTE).send({
        activityDesc: 'This is a nice activity',
        activityName: 'Test Activity',
        organizationId: 'afd195f1-cf17-480c-97aa-ee884fa00d24',
      });

      expect(res.status).toBe(400);
      expect(res.body).toStrictEqual({
        error: 'OrganizationDoesNotExist',
        additionalInfo: {},
      });
    });

    it('should not create activity type when name not unique for organization', async () => {
      const activityName = 'Test Activity';
      await prisma.activityType.create({
        data: {
          activityDesc: 'This is a nice activity',
          activityName,
          orgId: organization.id,
        },
      });

      const res = await supertest(app).post(ROUTE).send({
        activityDesc: 'This is a nice activity',
        activityName,
        organizationId: organization.id,
      });

      expect(res.status).toBe(400);
      expect(res.body).toStrictEqual({
        error: 'UnableToCreateResource',
        additionalInfo: {},
      });
    });

    it('should create activity if name already exists for other organization', async () => {
      const activityName = 'Test Activity';
      await prisma.activityType.create({
        data: {
          activityDesc: 'This is a nice activity',
          activityName,
          orgId: organization.id,
        },
      });

      const res = await supertest(app).post(ROUTE).send({
        activityDesc: 'This is a nice activity',
        activityName,
        organizationId: otherOrganization.id,
      });

      expect(res.status).toBe(201);
      expect(await prisma.activityType.count()).toBe(2);
    });
  });

  describe('getAllActivities', () => {
    it('should return all activities for a company', async () => {
      await prisma.activityType.create({
        data: {
          activityDesc: 'This is a nice activity',
          activityName: 'Some Activity',
          orgId: organization.id,
        },
      });

      const res = await supertest(app).get(`${ROUTE}/${organization.id}`);

      expect(res.status).toBe(200);
      expect(res.body).toStrictEqual([
        {
          id: expect.any(String),
          activityName: 'Some Activity',
          activityDesc: 'This is a nice activity',
          orgId: organization.id,
        },
      ]);
    });

    it('should not return activities for another company', async () => {
      await prisma.activityType.create({
        data: {
          activityDesc: 'This is a nice activity',
          activityName: 'Some Activity',
          orgId: organization.id,
        },
      });

      const res = await supertest(app).get(`${ROUTE}/${otherOrganization.id}`);

      expect(res.status).toBe(200);
      expect(res.body).toStrictEqual([]);
    });
  });
});
