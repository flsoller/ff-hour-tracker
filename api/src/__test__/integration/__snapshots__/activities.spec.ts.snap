// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activities controller createActivity should create an activity for an organization 1`] = `
Object {
  "activityDesc": "This is a nice activity",
  "activityName": "Test Activity",
  "id": Any<String>,
}
`;
